File: 백준/Silver/15649. N과 M （1）/N과 M （1）.py
import sys

input = sys.stdin.readline

N, M = map(int, input().split())
arr = []
v = [False] * (N + 1)


def dfs(n):
    if n == M:
        print(*arr)
        return
    for i in range(1, N + 1):
        if not v[i]:
            arr.append(i)
            v[i] = True
            dfs(n + 1)
            arr.pop()
            v[i] = False


dfs(0)

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
import java.io.*;

// Press Shift twice to open the Search Everywhere dialog and type `show whitespaces`,
// then press Enter. You can now see whitespace characters in your code.
public class Main {
    public static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static int[] arr;
    public static boolean[] visit;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        arr = new int[M];
        visit = new boolean[N];
        dfs(N, M, 0);
        bw.flush();
        bw.close();
    }

    public static void dfs(int N, int M, int depth) throws IOException {
        if (depth == M)
        {
            for (int val: arr)
                bw.write(val + " ");
            bw.write("\n");
            return;
        }
        for (int i = 0; i < N; i++){
            if (!visit[i]) {
                visit[i] = true;
                arr[depth] = i + 1;
                dfs(N, M, depth + 1);
                visit[i] = false;
            }
        }
    }
}